dockstore-client/src/main/java/io/dockstore/client/cli/YamlVerify.java

```diff
@@ -0,0 +1,209 @@
+
+        // Check Workflows
+        List<YamlWorkflow> workflows = dockstoreYaml12.getWorkflows();
+        if (workflows != null) {
+            for (YamlWorkflow workflow : workflows) {
+                List<String> filePaths = workflow.getTestParameterFiles();
+                filePaths.add(workflow.getPrimaryDescriptorPath());
+                missingFiles += filesExist(filePaths, basePath);
+            }
+        }
+        // Check Tools
+        List<YamlWorkflow> tools = dockstoreYaml12.getTools();
+        if (tools != null) {
+            for (YamlWorkflow tool : tools) {
+                List<String> filePaths = tool.getTestParameterFiles();
+                filePaths.add(tool.getPrimaryDescriptorPath());
+                missingFiles += filesExist(filePaths, basePath);
+            }
+        }
+
+        // Check service
+        Service12 service = dockstoreYaml12.getService();
+        if (service != null) {
+            missingFiles += filesExist(service.getFiles(), basePath);
+        }
+        if (!missingFiles.isBlank()) {
+            throw new ValidateYamlException("Your file structure has the following errors:\n" + missingFiles);
+        }
+    }
+
+    public static void dockstoreValidate(String path) throws ValidateYamlException {
+        Path workflowPath = Paths.get(path);
+
+        // Verify that the path is a valid directory
+        if (!Files.isDirectory(workflowPath)) {
+            throw new ValidateYamlException(ERROR_MESSAGE + workflowPath.toString() + " does not exist");
+        }
+
+        // Verify that .dockstore.yml exists
+        Path dockstoreYmlPath = Paths.get(path, DOCKSTOREYML);
+        if (!Files.exists(dockstoreYmlPath)) {
+            throw new ValidateYamlException(ERROR_MESSAGE + dockstoreYmlPath.toString() + " does not exist");
+        }
+
+        // Verify that .dockstore.yml is non-empty
+        File dockstoreYml = new File(dockstoreYmlPath.toString());
+        if (dockstoreYml.length() == 0) {
+            throw new ValidateYamlException(ERROR_MESSAGE + dockstoreYmlPath.toString() + " is empty");
+        }
+
+        // Verify that the Yaml is valid, however does not verify it is valid for dockstore
+        Yaml yaml = new Yaml();
+        Map<String, Object> data = null;
+        try {
+            data = yaml.load(Files.readString(dockstoreYmlPath));
+            out(dockstoreYmlPath + " is a valid yaml file");
+        } catch (Exception ex) {
+            throw new ValidateYamlException(ERROR_MESSAGE + dockstoreYmlPath.toString() + " is not a valid yaml file:\n" + ex.getMessage());
+        }
+
+        // Running validate first, as if readAsDockstoreYaml12 is run first and a parameter is incorrect it is difficult to understand
+        // Determine if any of the parameters names are incorrect (does not verify if all the parameters are there)
+        try {
+
+            DockstoreYamlHelper.validateDockstoreYamlProperties(Files.readString(dockstoreYmlPath));
```

Comment:
Minor nit, I'd just read the file just once and save the value (you read it again on line 180). And catch that specific exception, e.g., you're unable to read the file.

Reply:
Thanks for the feedback, I have fixed this.